File: java/153Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        for (int i = 0; i < nums.length; ++i) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            
            int j = i + 1, k = nums.length - 1;
            while (j < k) {
                if (-nums[i] > nums[j] + nums[k]) {
                    ++j;
                } else if (-nums[i] < nums[j] + nums[k]) {
                    --k;
                } else {
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    while (j < k && nums[j] == nums[j + 1])
                        ++j;
                    ++j;
                    while (j < k && nums[k] == nums[k - 1])
                        --k;
                    --k;
                }
            }
        }
        
        return res;
    }
}

File: java/153Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        C=collections.Counter(nums)
        finds,results,bisl=sorted(C),set([(0,0,0)]) if C[0]>=3 else set(),bisect.bisect_left
        for i,k1 in enumerate(finds):
            if C[k1]>1 and -2*k1 in C and k1:   results.add((k1,k1,-2*k1))
            if k1<0:
                for k2 in finds[bisl(finds,-finds[-1]-k1,i+1):bisl(finds,math.ceil(-k1/2),i+1)]:
                    if -(k1+k2) in C:results.add((k1,k2,-k1-k2))
        return results
